### DISCUSSION TITLE
Do you think the Svelte API / pattern of use could be improved?

### DISCUSSION THREAD

**Discussion by ispyinternet** [urql-graphql/urql#997] [Q&A] [answered]
I would argue that the use of context to save the Client reference is brittle and potentially confusing due to svelte requiring that you call this code at a very specific place/time in the component code.

```js
// component
<script>
import { queryString } from 'query.graphql.js'

const todosQuery = query({ query: queryString })

$: todos = todosQuery({ pause: true });
const execute = () => todosQuery().then();
</script>
```

I was trying to visualise a setup, where perhaps you construct your own client, and then use it to prepare all your queries / mutations etc. So the client exports 3 methods, query, mutate, and subscribe, that you could use similar to the gql tagging function on your .graphql schemas:

```js
// includes/urql.config.js 
import { SvelteClient } from '@urql/svelte;
const client =  new SvelteClient( { constructor_args })

export default client; 
export { client.query as q, client.mutate as m, client.subscribe as s }
```

```js
// graphql/queries.js
import { q } from 'urql.config.js'
export const todos = q`
  query {
        todos {
          id
          text
          complete
        }
      }`;
```

```js
// component
<script>
import { todos } from 'query.js'

const execute = () => todos.update({ new_args});
</script>

{#if $todos.fetching}
  Loading...
{:else if $todos.error}
  Oh no! {$todos.error.message}
{:else if !$todos.data}
  No data
{:else}
  <ul>
    {#each $todos.data.todos as todo}
      <li>{todo.text}</li>
    {/each}
  </ul>
{/if}

<button on:click={execute}>Execute</button>

**Comment by ispyinternet** [comment]
I might add that this has been thought out in the context of a larger sapper app

**Comment by kitten** [comment, marked as answer]
We opt not to wrap `urql`'s `Client` ever because that introduces confusion around what the client is and what its methods currently are. That being said our Svelte bindings are purely experimental and stand-ins as there are (imho) deep problems with Svelte's reactive model, to the point where most code just opts to use `await`, which isn't optimal.

What you're describing however is already possible. You can actually use `query` directly without instantiating it first into a separate variable for instance. The reason why this is done currently is precisely because of Svelte's shortcoming in how subscribables are handled

The problem is that whenever it changes it loses all its state and currently we've "hacked" around with a small trick. If you instantiate the query first and then use it separately with another call, we're able to actually preserve the state as we want to.

This may change though as we think about what the final API should look like. It's pretty likely we'll actually just drop this and "blame" Svelte's (probably intentional) limitations on why our bindings won't be as fully-featured as the ones for React/Preact 😅

**Reply by ispyinternet** [reply]
At the moment it is not possible to use the provided svelte query method unless it is called synchronously during component initialization as it makes a call out to getContext?

**Reply by ispyinternet** [reply]
> 
> The problem is that whenever it changes it loses all its state and currently we've "hacked" around with a small trick. If you instantiate the query first and then use it separately with another call, we're able to actually preserve the state as we want to.
> 
I'm a bit wet behind the ears, can you help me to understand what state you are trying to preserve and how to relate this issue to a real world scenario? Thanks

**Reply by kitten** [reply]
Ok, so basically when we have a change in inputs in the Svelte component then Svelte will compile this so that when that change happens a new stream is created and subscribed to. Why is that an issue?

Multiple parts of the reactive API in React/Preact rely on some state being preserved, i.e. a full input instead of a recreation. This happens because you may change variables but are still interested in the previous result. This is a limitation we wanted to experiment with. So if you call `query` and then you call the function it returns it tried to preserve some of this underlying state.

If you don't call this function and just use it as a store, then it won't preserve the state

**Reply by ispyinternet** [reply]
State as in .fetching .error .data? Essentially data might revert to ‘empty’ while a new query is sent over the wire? Perhaps other effects I’m not clear on yet?

**Reply by ispyinternet** [reply]
I can't help thinking that in svelte-land if urql were to expose the query (operations) as readable stores that included a retry (perhaps call it bind) method. then the original 'subscribable' wouldn't be destroyed and the api would feel a bit more natural / svelteish.

We could setup our reactive bindings, without destroying the observable like this:

```js
// component
import { todos } from 'queries.js'

let nextToken;
let maxItems = 10;

$: $todos.bind( { nextToken, maxItems }) // bind the query to our input variables

</script>
{#if $todos.fetching}
  Loading...
{:else if $todos.error}
  Oh no! {$todos.error.message}
{:else if !$todos.data}
  No data
{:else}
  <ul>
    {#each $todos.data.todos as todo}
      <li>{todo.text}</li>
    {/each}
  </ul>
<button on:click="{() => nextToken = $todos.data.nextToken}">Next Page</button>
{/if}
```

**Reply by ispyinternet** [reply]
@kitten r̶e̶g̶ ̶`̶c̶l̶i̶e̶n̶t̶.̶q̶u̶e̶r̶y̶`̶ ̶a̶n̶d̶ ̶`̶c̶l̶i̶e̶n̶t̶.̶e̶x̶e̶c̶u̶t̶e̶Q̶u̶e̶r̶y̶`̶,̶ ̶I̶ ̶s̶e̶e̶ ̶i̶n̶ ̶t̶h̶e̶ ̶r̶e̶a̶c̶t̶ ̶p̶a̶c̶k̶a̶g̶e̶ ̶t̶h̶e̶ ̶`̶e̶x̶e̶c̶u̶t̶e̶Q̶u̶e̶r̶y̶`̶ ̶w̶i̶l̶l̶ ̶u̶s̶e̶ ̶a̶ ̶m̶e̶m̶o̶i̶z̶e̶d̶ ̶t̶h̶e̶ ̶G̶r̶a̶p̶h̶Q̶L̶ ̶r̶e̶q̶u̶e̶s̶t̶,̶ ̶I̶'̶m̶ ̶g̶u̶s̶s̶i̶n̶g̶ ̶b̶y̶ ̶u̶s̶i̶n̶g̶ ̶`̶c̶l̶i̶e̶n̶t̶.̶q̶u̶e̶r̶y̶`̶ ̶y̶o̶u̶ ̶l̶o̶s̶e̶ ̶t̶h̶i̶s̶?̶ ̶h̶o̶w̶e̶v̶e̶r̶ ̶d̶o̶e̶s̶ ̶t̶h̶e̶ ̶c̶a̶c̶h̶e̶ ̶l̶a̶y̶e̶r̶ ̶k̶i̶n̶d̶ ̶o̶f̶ ̶d̶o̶u̶b̶l̶e̶ ̶d̶o̶w̶n̶ ̶o̶n̶ ̶t̶h̶i̶s̶ ̶o̶p̶t̶i̶m̶i̶s̶a̶t̶i̶o̶n̶ ̶o̶f̶ ̶r̶e̶q̶u̶e̶s̶t̶s̶? obviously one use a memoized query the other doesn't - ultimately I expect you would prefer to achieve this with svelte.

I think the svelte package could be improved and I'm having a play around with these patterns. Did you see the above pattern I suggested for binding the variables?

If I thought your team were look at this I would hold off, otherwise I would like to try and help improve the svelte package (in which case where would be the best place to ask more questions?)

Another question I had, in react `useMutation` you check for mounted

`if (isMounted.current) { setState...`

Is there a reason you don't perform this check in useQuery?

**Reply by kitten** [reply]
That check isn't performed in `useQuery` because it's already guaranteed by `useSource`: https://github.com/FormidableLabs/urql/blob/8c74924c1d49588f265cbc3776e98fbff4995199/packages/react-urql/src/hooks/useSource.ts#L46

**Reply by ispyinternet** [reply]
what is the point of `useMemo` if you don't pass any arguments in the second arg / array? (sorry don't know react!)

https://github.com/FormidableLabs/urql/blob/8c74924c1d49588f265cbc3776e98fbff4995199/packages/react-urql/src/hooks/useSource.ts#L55

**Reply by kitten** [reply]
It is passing “something” 😛 the dependencies array is empty, which means that we avoid changing the returned value in React’s hooks state, i.e. telling React that it shouldn’t recompute it.

But it’s still a memoisation and not hard state so that during hot updates and other cases this memoisation is still broken.

**Reply by ispyinternet** [reply]
Ah! I miss read:

> If no array is provided, a new value will be computed on every render

You are providing an empty array!

**Reply by frederikhors** [reply]
@ispyinternet After some months and projects with Svelte 3 React looks like a dinosaurus!!! I'm with you!

**Reply by ispyinternet** [reply]
@kitten btw, these 'memoizations' (essentially all those react hooks!) - they are just interested to know if the value has changed from the previous - do any of them need to implement a history/cache store? I don't think so? - in fact do the react hooks even provide that?

**Reply by kitten** [reply]
@ispyinternet No, this isn't about memoization and a history per se, instead it's about having a coherent change from one state to the next. This stream here shows the state changes very clearly, `scan` being like a reducer over time: https://github.com/FormidableLabs/urql/blob/8c74924c1d49588f265cbc3776e98fbff4995199/packages/react-urql/src/hooks/useQuery.ts#L69-L96

@frederikhors I wouldn't say that at all. In my opinion, Svelte's reactivity and state models are deeply insufficient and flawed compared to component-wide state via hooks. Neither is satisfactory and in fact in both models there are state changes and events that we cannot observe; with React we cannot observe the exact time a render cycle is interrupted and in Svelte we cannot react to and observe changes over time at all cleanly. The latter is in my opinion a lot more severe

**Reply by ispyinternet** [reply]
Thanks @kitten. Another question, not sure how you handle this, if it’s just left to user land, but if your binding query params to template vars, depending on how those input changes are emitted (eg keyups on an input - I assume the user should implement the pause here), but do you do any debouncing / atomising of the template variable changes to prevent firing off a splatter of similar requests? Hope that question’s not too dumb in the context of this library!?🤷 

Yes that code snippet you sent, just getting my head round it, learning wonka!

**Reply by ispyinternet** [reply]
Hi @kitten ok so... I hope what I have done will amount to something! I've had to learn typescript along the way - which I'm totally sold on, I would NOT have been able to do this without the IDE type errors!

So the theory - svelte stores maintain the references in the same way your react hooks do and any changes to the underlying stores will propagate to the final query through the use of derived stores.

We have:

*Request*
 - provides `GraphQLRequest`
 - changed by: `query`, `variables`
 - Code: https://gist.github.com/ispyinternet/12e2311a74f57d7f83e7b17ee0c28ea4
 - Notes: I've asked a question in the source code here - do I need to check the request key as you do in react?

*Context*
 - provides `OperationContext`
 - changed by: `pollInterval`, `requestPolicy` (*more dependencies can be added*)
 - Code: https://gist.github.com/ispyinternet/ad28499a98350227f819196271f64773

*Source*
 - provides `Source<OperationResult<T>>` 
 - changed by: `request`, `context`
 - Code: https://gist.github.com/ispyinternet/8bd7b1eb307f0a6ead3fe8ca656c8c62

*Query*
 - provides: `Source<OperationResult<T>`
 - changed by: `source`
 - Code: https://gist.github.com/ispyinternet/05372d2963b301108c43a9447c578e43
 - Notes: Orchestrates changes to the stores, sets intial state and manages pausing, allows us to attach reactivity in svelte template using for example: `$: $todos.bind( { variables: { nextToken, maxItems }})`

The reason I asked the question about debouncing / atomising of underlying changes, is because, in these examples, where changes to things like query, variables and context are made synchronously, they will trigger multiple updates on the derived store - I can't reason what happens in this scenario, wonka will just quickly recreate the request and unsubscribe to the previous / kill any network request - do I need to add some logic here to prevent multiple network requests being created (debounce) - do you see / understand what I am trying to convey?

It would be great if you could look over them, I'm not 100% sure about the last one `query.ts` - have I constructed the wonka stream correctly? I'm getting a type error I can't fix:

<img width="450" alt="Screenshot 2020-09-30 at 16 32 45" src="https://user-images.githubusercontent.com/1741823/94706891-aff53000-033a-11eb-854d-930302649290.png">

Cheers!

**Reply by ispyinternet** [reply]
Just realise that `reexecuteQuery` will not necessarily re-execute the query (only if something in context changes)! 

I will look into this - any thoughts? I could trick the context store into changing with a pseudo variable (incrementer or something)

**Reply by kitten** [reply]
The comment below could be of interest; since we have a new stable API we'll ship for Svelte.

> do I need to add some logic here to prevent multiple network requests being created (debounce) - do you see / understand what I am trying to convey?

No, the cancellation is pretty transparent and cheap. We'd choose not to interfere with debouncing or other time-related logic and would instead opt for predictability. It's typically not a problem if a query rapidly changes. If it is identical to the last one then the `dedupExchange` will also just let the first one through and any further cancellation isn't necessary. If the previous operation is cancelled this just flows through to the `fetchExchange`.

**Comment by kitten** [comment]
So to bring this thread to a more satisfying conclusion; I've fast-tracked the development of a new Svelte bindings API in https://github.com/FormidableLabs/urql/pull/1016.

This API should protect the users more rigorously from misusing the reactivity model that we had before, and should appear much more intuitive. It also irons out several bugs that we found while experimenting with different APIs.

This is still not 100% perfect, but it's the closest we can get to an idiomatic Svelte API that also feels right for urql and GraphQL.

**Reply by ispyinternet** [reply]
I did ask if your team were looking into this and I would have held off!? anyway...

**Reply by ispyinternet** [reply]
think I prefer my suggested bind api though!?

**Reply by kitten** [reply]
We're still figuring out discussions but we did indeed forget to link #1016 to this thread; that being said, I appreciate the reflection on some of the problems we've been having but we do schedule work via issues to prevent conflicts 😅 ❤️

**Reply by ispyinternet** [reply]
I'll forgive if you could take the time to review / critique my code!? (would it have worked!?) It was a good exercise - especially discovering how useful typescript is!

Reverse engineering the react code was a brain teaser though!

Thanks in advance! 😀

**Reply by kitten** [reply]
I think it would! I'm not quite sure; Svelte code is still a little unpredictable to me. So I think there are some places where I'm not sure whether the changes are flowing down correctly (mainly `context` in `query`) but on the other hand. So would `bind` be a manual call or is it a Svelte compiler target that I don't know about?

**Reply by kitten** [reply]
I think one thing that I've noticed is that this wouldn't really subscribe and this is also what I'm mostly unsure about https://gist.github.com/ispyinternet/05372d2963b301108c43a9447c578e43#file-query-ts-L53-L83

**Reply by ispyinternet** [reply]
so in a svelte template if you prefix an expression with `$:` it becomes reactive to any of the variables in the expression so:

```js
export let var = 1;
$: console.log(var) // when ever var changes (either from outside or template binding, console logs)
```
so in urql example:
```js
$: todos.bind( { variables: { nextToken, maxItems }}) // when ever nextToken or maxItems change bind is called (which will set value on request store - which may trigger update on source - which will trigger update on query - I think!?
// I think this should just be `todos.bind` `$todos` being the value of `subscribe`
```

**Reply by ispyinternet** [reply]
> I think one thing that I've noticed is that this wouldn't really subscribe and this is also what I'm mostly unsure about

yh probaly the bit I was unsure about.

the query store if you prefix with `$` in the template should return the value of subscribe.

see this proof of concept, cascading changes to derived store, plus noop (i.e. retain references/state) if updating to same value: 

https://svelte.dev/repl/a67c9d37aee348d988e8c30f60a139d9?version=3.28.0

**Reply by kitten** [reply]
Yea, I mean, the reactivity is fine, but what I don't see is `bind` being the right approach here? So basically why I went with the `operationStore` in the upcoming implementation is because that means that we don't have to pass mutable variables, but we pass a container that also owns the changes, which means there's only one source of changes to worry about

**Reply by ispyinternet** [reply]
That I will have to go away and think about! Cheers!

**Reply by ispyinternet** [reply]
> This is still not 100% perfect, but it's the closest we can get to an idiomatic Svelte API that also feels right for urql and GraphQL.

@kitten why?

**Reply by ispyinternet** [reply]
> Yea, I mean, the reactivity is fine, but what I don't see is `bind` being the right approach here? So basically why I went with the `operationStore` in the upcoming implementation is because that means that we don't have to pass mutable variables, but we pass a container that also owns the changes, which means there's only one source of changes to worry about

Yh, tbf, I think bind was the wrong choice of word, nothing was binding, it was just updating the values in the underlying stores.

Essentially similar solutions, in your solution, I',m pretty sure you can write:

```js
$: $todos.variables.skip = skip; // whenever skip changes either from external or within, the query gets triggered.
```

For posterity I got mine working with: https://gist.github.com/ispyinternet/05372d2963b301108c43a9447c578e43 !

I will argue my api is a little more succinct as you don't need to expose the operationStore to the end user, `query` sets all the stores up for you.

```html
<script>
import {  query } from '@urql/svelte';
import { todos as query } from './graphyql/todos';

let pause = true;
const todos = query({ query, pause })

$: todos.execute({ pause }); 
</script>

<button on:click="{() => pause = false}">Run Query</button>

{#if $todos.fetching} etc...
```

Anyway I only got query working, and I don't have the inclination to work out mutations or subscriptions!

Thanks for the time and great work!
